Add check room availabity

pymanceva opened the pull request:
Карточки: https://nkiblyk.kaiten.ru/space/167883/card/27548967?filter=eyJrZXkiOiJhbmQiLCJ2YWx1ZSI6W3sia2V5IjoiYW5kIiwidmFsdWUiOlt7ImNvbXBhcmlzb24iOiJlcSIsImtleSI6InJlc3BvbnNpYmxlIiwidmFsdWUiOjQ5NDU3MH1dfV19

https://nkiblyk.kaiten.ru/space/167883/card/27548993?filter=eyJrZXkiOiJhbmQiLCJ2YWx1ZSI6W3sia2V5IjoiYW5kIiwidmFsdWUiOlt7ImNvbXBhcmlzb24iOiJlcSIsImtleSI6InJlc3BvbnNpYmxlIiwidmFsdWUiOjQ5NDU3MH1dfV19

US: https://docs.google.com/document/d/1yG7TwO6sCLbJMLEOxF_YmKiB2seioYF4fFHz6Uc-weU/edit
US: https://nkiblyk.kaiten.ru/documents/d/7a1c8a2d-b776-45bc-adcf-0a7091b5a9bc

Commit:
feat: добавить проверку доступности номеров перед бронью

Commit:
feat: добавить выдачу пересекающихся бронирований

Commit:
# Conflicts:
#	src/main/java/ru/modgy/room/dto/mapper/RoomMapper.java

Commit:
# Conflicts:
#	pom.xml
#	src/main/java/ru/modgy/booking/service/BookingServiceImpl.java
#	src/main/java/ru/modgy/pet/controller/PetController.java
#	src/main/java/ru/modgy/pet/service/PetServiceImpl.java
#	src/main/java/ru/modgy/room/dto/mapper/RoomMapper.java
#	src/main/java/ru/modgy/room/service/RoomServiceImpl.java
#	src/test/java/ru/modgy/room/service/RoomServiceImplTest.java

Commit:
feat: добавить выдачу блокирующих бронирований для номера

Commit:
fix: поправить тесты

Commit:
fix: поправить тесты

Commit:
feat: тесты для проверок доступности номеров

Commit:
fix: дополнить swagger

RodionovRL commented on pet-hotel-versions/CHANGELOG.md:
```diff
@@ -35,4 +35,19 @@ Api: *<https://github.com/pymanceva/pet-hotel/blob/develop/Specification/pet-hot
+**0.0.11: Добавлен UtilityService для получения сущностей всех типов и проверки прав доступа.**
+Карточка задачи: *<https://nkiblyk.kaiten.ru/space/167883/card/27549007?filter=eyJrZXkiOiJhbmQiLCJ2YWx1ZSI6W3sia2V5IjoiYW5kIiwidmFsdWUiOlt7ImNvbXBhcmlzb24iOiJlcSIsImtleSI6InRhZyIsInZhbHVlIjo0OTA5MDF9XX1dfQ>*
+
+**0.0.12: Добавлена проверка доступности номеров, поиск блокирующих и пересекающихся бронирований 
```
Номер версии не угадала =)

pymanceva replied:
на момент ПР ты не мержил, естественно, я ставила номер по порядку. Мы ранее уже обсуждали это, мне так удобнее, потом перед мержем правлю версию

RodionovRL commented on pet-hotel-versions/docker-compose.yaml:
```diff
@@ -2,7 +2,7 @@ version: '3.8'
 services:
   server:
-    build: .
+    build: modgy-back-0.0.9
```
почему именно эта папка?

pymanceva replied:
я не трогала докер, это то что прилетело с чужих правок. Честно говоря, уже вообще не понимаю, как гитхаб отражает изменения. Это не мои исправления

pymanceva replied:
единственное - подвигала папки. Мы как-то обсуждали, что у нас докерфайл в каждую папку идет, и changelog, и т.д. Я их вынесла выше, в в папках версий оставила только jar. Но внутрянке в докерфайле и докеркомпоузе не трогала

RodionovRL replied:
Оставляй как есть.

RodionovRL commented on pom.xml:
```diff
@@ -10,7 +10,7 @@
 	<groupId>ru.dogudacha</groupId>
 	<artifactId>Modgy</artifactId>
-	<version>0.0.8-SNAPSHOT</version>
+	<version>0.0.12-SNAPSHOT</version>
```
0.0.13

pymanceva replied:
про версию выше написала

RodionovRL commented on specification/pet-hotel-rooms-spec.yaml:
```diff
@@ -15,7 +17,7 @@ paths:
         - room-controller
       operationId: getAllRooms
       summary: Получение списка всех номеров
```
Что-то не нашёл в юзерстори запрос всех номеров.

pymanceva replied:
это старый код, написанный ранее, прошедший ревью ранее, еще до прихода Наташи и написания юзерстори.

RodionovRL commented on specification/pet-hotel-rooms-spec.yaml:
```diff
@@ -15,7 +17,7 @@ paths:
       operationId: getAllRooms
       summary: Получение списка всех номеров
-      description: Получение списка всех номеров. Доступно только пользователями с ролью ROLE_BOSS или ROLE_ADMIN.
+      description: Получение списка всех номеров. Доступно только пользователями с ролью ROLE_BOSS, ROLE_ADMIN, ROLE_FINANCIAL.
```
Финансист тоже может просматривать список всех номеров? Не нашёл в Юзерстори

pymanceva replied:
это старый код, написанный ранее, прошедший ревью ранее, еще до прихода Наташи и написания юзерстори. На тот момент мы обсуждали доступ к инфо о номерах у финансиста в том числе. Чтобы он имел возможность смотреть, сколько у гостиницы их всего, каких категорий, какая стандарт цена и т.п. Для фин прогнозов и тп

RodionovRL replied:
Давай попросим Наташу подправить юзерстори, наверное так будет правильно?

pymanceva replied:
я не против, но если будем править старый код - все равно в отдельную задачу, не в рамках нынешней. Так что оставляю resolved, и далее видно будет

RodionovRL commented on specification/pet-hotel-rooms-spec.yaml:
```diff
@@ -167,6 +171,8 @@ paths:
         - room-controller
       operationId: getRoomById
+      summary: Получение номера по его ID
+      description: Получение номера по его ID. Доступно только пользователями с ролью ROLE_BOSS, ROLE_ADMIN, ROLE_FINANCIAL.
```
Финансист может просматривать номер? Не нашёл в Юзерстори

pymanceva replied:
отвечала выше

RodionovRL commented on src/main/java/ru/modgy/booking/service/BookingService.java:
Javadoc - красотища =)

RodionovRL commented on src/main/java/ru/modgy/booking/service/BookingServiceImpl.java:
```diff
@@ -142,16 +137,17 @@ public BookingDto updateBooking(Long userId, Long bookingId, UpdateBookingDto up
         if (updateBookingDto.getRoomId() != null) {
-            newBooking.setRoom(findRoomById(updateBookingDto.getRoomId()));
+            Room room = entityService.getRoomIfExists(updateBookingDto.getRoomId());
+            checkRoom(room);
```
Здесь вызывается метод, в котором проверяется isAvalible с текстом ошибки:
"Can't add booking for hidden room"
при этом мы не добавляем, а апдейтим наше бронирование.

Возможным решением видится передача в метод не только объекта Room, но и текст, который будет выдан при ошибке.

RodionovRL replied:
ЗанудаMode on 
было бы круто, если бы это вылезло на тестах.
ЗанудаMode off

pymanceva replied:
При обновлении бронирования также обновленная инфа добавляется в базу. Да, она вымещает старую, но мы даже метод используем тот же, что и при создании нового объекта - save. Кроме того, в рамках возможного выплывания этой ошибки - мы меняем номер в брони. То есть для нового номера - это попытка добавить бронирование. Ранее у номера этой брони не было. Так что формально тоже текст подходит

RodionovRL replied:
Ну название метода значения не имеет, конечно. Здесь скорее про то, что, если мы обновляем что-то, то странно в ответ получить, что не удалось создать(мы ведь и не создавали ничего, мы уже созданное редактируем)

RodionovRL commented on src/main/java/ru/modgy/owner/dto/mapper/OwnerMapper.java:
Вангую здесь конфликт при меже =)

pymanceva replied:
да, но конфликты есть всегда, выбрать версию девелопа не проблема. А без этого не билдился проект в принципе

RodionovRL commented on src/main/java/ru/modgy/pet/controller/PetController.java:
```diff
@@ -17,36 +21,57 @@
         return petService.getPetById(requesterId, petId);
     }
 
+    @GetMapping("/search")
```
Не понял как в этом ПР оказался поиск питомцев, ну да ладно =)

pymanceva replied:
я вообще хз, реально. Иногда неведомым образов влетают чужие блоки, и в большом количестве. В питомцах тогда все юзеры засветились, помню. Тут питомцы

RodionovRL replied:
Чудные дела :) 
Ну да шут с ним :)

RodionovRL commented on src/main/java/ru/modgy/pet/dto/NewPetDto.java:
```diff
@@ -15,7 +15,7 @@
 @NoArgsConstructor
 public class NewPetDto {
-//    @NotNull(message = "Field: owner. Error: must not be blank.")
+//todo    @NotNull(message = "Field: owner. Error: must not be blank.")
```
Красотища =)

pymanceva replied:
это даже не мое))) это Кристинины питомцы)

RodionovRL commented on src/main/java/ru/modgy/pet/model/Pet.java:
```diff
@@ -18,7 +20,7 @@ public class Pet {
     @Column(name = "id_pets")
     private long id;
-    //    @ManyToOne()
+//        @ManyToOne()
```
TODO?

pymanceva replied:
Снова не мое, питомцев пишет Кристина

RodionovRL commented on src/main/java/ru/modgy/pet/model/Pet.java:
```diff
@@ -134,6 +136,9 @@ public class Pet {
     private String medicineRegimen;
     @Column(name = "additional_data_pets", nullable = true, length = 1000)
     private String additionalData;
+    @CreationTimestamp
```
Элегантно.

RodionovRL commented on src/main/java/ru/modgy/pet/model/TypeOfPet.java:
```diff
@@ -1,9 +1,7 @@
-
     CAT,
-
+    DOG,
```
сместили пёселя по алфовиту?

pymanceva replied:
аналогично

RodionovRL commented on src/main/java/ru/modgy/pet/repository/PetRepository.java:
```diff
@@ -6,7 +6,7 @@
 
-public interface PetRepository extends JpaRepository<Pet, Long> {
+public interface PetRepository extends JpaRepository<Pet, Long>, SearchPetRepository {
 //    Pet findByOwnerAndName(long ownerId, String namePet);
```
TODO

RodionovRL commented on src/main/java/ru/modgy/pet/service/PetServiceImpl.java:
```diff
@@ -247,27 +247,51 @@ public void deletePetById(Long requesterId, Long petId) {
-        return userRepository.findById(userId).orElseThrow(() ->
-                new NotFoundException(String.format("User with id = %d not found", userId)));
-    }
     //метод проверки наличия хозяина питомца, будет дописан после добавления сущности оунеров
```
todo

RodionovRL commented on src/main/java/ru/modgy/room/RoomController.java:
```diff
@@ -10,70 +10,92 @@
+    public RoomDto getRoomById(@RequestHeader(UtilityService.REQUESTER_ID_HEADER) Long requesterId,
                                @PathVariable("id") Long roomId) {
         log.info("RoomController: GET/getRoomById, requesterId={}, roomId={}", requesterId, roomId);
+        utilityService.checkBossAdminFinancialAccess(requesterId);
```
Вопрос про финансиста аналогичен вопросу в сваггере

pymanceva replied:
ранее ответила, да

RodionovRL commented on src/main/java/ru/modgy/room/RoomController.java:
```diff
@@ -10,70 +10,92 @@
+    public Collection<RoomDto> getAllRooms(@RequestHeader(UtilityService.REQUESTER_ID_HEADER) Long requesterId,
                                            @Param("isVisible") Boolean isVisible) {
         log.info("RoomController: GET/getAllRooms, requesterId={}", requesterId);
+        utilityService.checkBossAdminFinancialAccess(requesterId);
```
вопрос про финансиста

RodionovRL commented on src/main/java/ru/modgy/room/RoomController.java:
```diff
@@ -10,70 +10,92 @@
+                                                            @PathVariable("catId") Long catId,
+                                                            @Param("checkInDate") LocalDate checkInDate,
+                                                            @Param("checkOutDate") LocalDate checkOutDate) {
+        log.info("RoomController: GET/getAvailableRoomsByCategoryInDates, " +
```
Валидировать параметры не будем?

pymanceva replied:
валидировать что именно? Ранее в контроллерах также ничего не валидировали, чисто required уточняли, если вдруг он false.

RodionovRL replied:
Ну если разобраться, то всегда именно в контроллере всё и валидируется. 
(Аннотации над полями во входящих dto)
Здесь можно даты валидировать, чтобы они были в правильном порядке и диапазоне, например.

pymanceva replied:
диапазон дат я затрудняюсь установить. Какие критерии взять? Просто из головы решить, что допустим 1900 год - вряд ли? Но где тогда грань? И про будущее также.
В сервисе я проверяю, чтобы дата заезда не была позже даты выезда, считаю так и нужно оставить, контроллер проверил наличие дат, ну и в целом формат, что это дата, а не что-то еще. А то, что они должны определенным образом между собой соотноситься - это же логика, это в сервис

RodionovRL replied:
Ну чтобы зря в сервис не ходить проверить, что даты не из прошлого, например. Мы ведь здесь проверяем доступность номера. Зачем нам эти данные из прошлого?

pymanceva replied:
Тут я уже идейно не согласна. Я могу захотеть добавить старую бронь, ы прошлом, она добавляться будет так же как и будущая, с проверкой доступности. Это может потребоваться для исправления ошибки (случайно удалили), или для корректной статистики (например, внесли старые брони по клиенту, хотим видеть исчерпывающую историю бронирований). Короче, даты в прошлом - ок

pymanceva replied:
Кажется, мы это в таком же ключе обсуждали при планировании ограничений в БД. И тоже тогда решили, что и задним числом всё вносить можно

RodionovRL replied:
Ну ок

RodionovRL commented on src/main/java/ru/modgy/room/category/service/CategoryServiceImpl.java:
```diff
@@ -85,33 +75,12 @@ public Collection<CategoryDto> getAllCategories(Long userId) {
         log.info("CategoryService: deleteCategoryById, userId={}, catId={}", userId, catId);
     }
-
-    public Category findCategoryById(Long id) {
```
Любопытно, что метод был публичный

RodionovRL commented on src/main/java/ru/modgy/room/dto/mapper/RoomMapper.java:
```diff
@@ -3,21 +3,20 @@
 
     @Mapping(target = "id", ignore = true)
-    @Mapping(target = "category", ignore = true)
+    @Mapping(source = "roomDto.categoryDto", target = "category")
```
Тоже красиво.

RodionovRL commented on src/main/java/ru/modgy/room/service/RoomService.java:
м-м-м-м, JavaDoc  
![image](https://github.com/pymanceva/pet-hotel/assets/114774699/d7861105-d339-41ca-85c0-3f2e512c4be6)

RodionovRL commented on src/main/java/ru/modgy/user/controller/UserController.java:
Здесь будет конфликт, прошу мою версию считать верной.

RodionovRL commented on src/main/java/ru/modgy/user/service/UserServiceImpl.java:
Здесь тоже может быть конфликт. Я уже проверял работу этого сервиса, когда писал оунеров.
Странно, это не в девелопе (только что понял это).

pymanceva replied:
ты смержил оунеров после ПР, так что тут многое из твоей ветки отсутствует. Повторюсь - конфликты есть в 99% случаев. Если будут сомнения при мерже - я приду с ними в чат. В остальном не понимаю сути комментариев)))

RodionovRL replied:
Теперь понял :)

RodionovRL commented on src/main/java/ru/modgy/utility/EntityService.java:
Обратно к вопросу о том, что утилитарки нету в develop до того, как случился ПР.

pymanceva replied:
Есть, я же сначала ту задачу до конца добила и смержила, потом еще все тут правила под утилитарку, потому что код был начат без него. На момент ПР утилитарка была в девелопе. Но каким боком это идет в текущий ПР - я не могу прокомментировать, не понимаю.

RodionovRL replied:
Ага, понял :)

RodionovRL commented on src/main/java/ru/modgy/utility/UtilityService.java:
Обратно к утилитарному классу.

RodionovRL commented on src/test/java/ru/modgy/booking/BookingControllerTest.java:
```diff
@@ -196,4 +204,67 @@ void deleteBookingById() {
+                        .header(requesterHeader, requesterId)
+                        .accept(MediaType.ALL_VALUE)
+                        .param("checkInDate", "2024-01-01" )
+                        .param("checkOutDate", "2024-01-02"))
```
Почему дата в формате ГГГГ.ММ.ДД?

RodionovRL commented on src/main/java/ru/modgy/booking/service/BookingServiceImpl.java:
```diff
@@ -160,7 +156,7 @@ public BookingDto updateBooking(Long userId, Long bookingId, UpdateBookingDto up
             newBooking.setStatus(StatusBooking.STATUS_CONFIRMED);
```
А чем-то обусловлено использование жёстких & и | вместо && и ||? 
Если нет, но надо заменить везде.

RodionovRL commented on src/main/java/ru/modgy/booking/service/BookingServiceImpl.java:
```diff
@@ -184,34 +178,59 @@ public void deleteBookingById(Long userId, Long bookingId) {
+            throw new ConflictException(String.format("Room with id=%d is not available for current dates", roomId));
         }
     }
 
```
238 строка:
```
 if (type.equals(TypesBooking.TYPE_CLOSING)) {
            if (reason == null) {
```

можно написать симпатичнее с использованием &&

pymanceva replied:
Можно вообще все переписать))) Это старый код, он даже не редактировался и по ошибке в текущий ПР не влетел. Если ты полагаешь, что эти правки необходимы - давай задачу в бэклог, в какой-то момент доберемся. В рамках текущей ветки это будет противоречием твоим же правилам, про то, чтобы на все отдельные задачи были

RodionovRL replied:
Да, значит надо карточку создать.

RodionovRL commented on src/test/java/ru/modgy/booking/BookingControllerTest.java:
```diff
@@ -31,14 +34,17 @@
 
 @WebMvcTest(controllers = BookingController.class)
 public class BookingControllerTest {
```
здесь public не нужен, как и в остальных тестах

RodionovRL reviewed (changes requested):
Отличная работа. Мы плавно движемся к нашей цели =)
Нужно внести немного правок. И учесть некоторые нюансы в связи с перспективой возникновения конфликтов при мерже.
Я не кидал запросы, принял за данность, что в этом плане всё работает.

Commit:
fix: заменить формат дат

Commit:
# Conflicts:
#	pet-hotel-versions/CHANGELOG.md
#	pet-hotel-versions/modgy-back-0.0.12/Modgy-0.0.12-SNAPSHOT.jar
#	specification/pet-hotel-pets-spec.yaml
#	specification/pet-hotel-rooms-spec.yaml
#	src/main/java/ru/modgy/booking/controller/BookingController.java
#	src/main/java/ru/modgy/booking/service/BookingServiceImpl.java
#	src/main/java/ru/modgy/owner/dto/mapper/OwnerMapper.java
#	src/main/java/ru/modgy/room/dto/mapper/RoomMapper.java
#	src/main/java/ru/modgy/room/service/RoomServiceImpl.java
#	src/main/java/ru/modgy/user/controller/UserController.java
#	src/main/java/ru/modgy/user/service/UserServiceImpl.java
#	src/main/java/ru/modgy/utility/EntityService.java
#	src/main/java/ru/modgy/utility/UtilityService.java
#	src/test/java/ru/modgy/booking/service/BookingServiceImplTest.java
#	src/test/java/ru/modgy/room/category/service/CategoryServiceImplTest.java
#	src/test/java/ru/modgy/room/service/RoomServiceImplTest.java
#	src/test/java/ru/modgy/user/service/UserServiceImplTest.java
#	src/test/java/ru/modgy/utility/EntityServiceTest.java
#	src/test/java/ru/modgy/utility/UtilityServiceTest.java

Commit:
feat: добавить версию, разрешение конфликтов

Commit:
fix: изменить текст ошибки в checkRoom

Commit:
fix: изменить CHANGELOG

pymanceva committed:
Merge branch 'develop' into add-check-room-availabity
